core/integration/storagetest/mutation_logs_reader.go

```diff
@@ -0,0 +1,177 @@
+
+// RunMutationLogsReaderTests runs all the tests against the provided storage implementation.
+func RunMutationLogsReaderTests(t *testing.T, factory logsRWFactory) {
+	ctx := context.Background()
+	b := &mutationLogsReaderTests{}
+	type TestFunc func(ctx context.Context, t *testing.T, f logsRWFactory)
+	for name, f := range map[string]TestFunc{
+		// TODO(gbelvin): Discover test methods via reflection.
+		"TestHighWatermark": b.TestHighWatermark,
+	} {
+		t.Run(name, func(t *testing.T) { f(ctx, t, factory) })
+	}
+}
+
+type mutationLogsReaderTests struct{}
+
+// TestHighWatermark ensures that reads respect the low inclusive, high exclusive API.
+// nolint:gocyclo
+func (mutationLogsReaderTests) TestHighWatermark(ctx context.Context, t *testing.T, newForTest logsRWFactory) {
+	directoryID := "TestHighWatermark"
+	logID := int64(1)
+	m, done := newForTest(ctx, t, directoryID, logID)
+	defer done(ctx)
+
+	// Setup the test by writing 10 items to the mutation log and
+	// collecting the reported high water mark after each write.
+	sent := []time.Time{} // Timesttamps that Send reported.
+	hwm := []time.Time{}  // High water marks collected after each Send.
+	maxIndex := 9
+	for i := 0; i <= maxIndex; i++ {
+		ts, err := m.Send(ctx, directoryID, logID,
+			&pb.EntryUpdate{Mutation: &pb.SignedEntry{Entry: []byte{byte(i)}}})
+		if err != nil {
+			t.Fatalf("Send(%v): %v", logID, err)
+		}
+		count, wm, err := m.HighWatermark(ctx, directoryID, logID, minWatermark, 100 /*batchSize*/)
+		if err != nil {
+			t.Fatalf("HighWatermark(): %v", err)
+		}
+		if want := int32(i) + 1; count != want {
+			t.Fatalf("HighWatermark(): count %v, want %v", count, want)
+		}
+		sent = append(sent, ts)
+		hwm = append(hwm, wm)
+	}
+
+	arbitraryTime := time.Date(1, 2, 3, 4, 5, 6, 7, time.UTC)
+	// Tests that query HighWatermarks with varying parameters and validate results directly.
+	for _, tc := range []struct {
+		desc  string
+		start time.Time
+		batch int32
+		want  time.Time
+	}{
+		// Verify that high watermarks preserves the starting time when batch size is 0.
+		{desc: "preserve start batch 0", start: arbitraryTime, batch: 0, want: arbitraryTime},
+		// Verify that high watermarks preserves the starting time when there are no rows in the result.
+		{desc: "preserve start rows 0", start: sent[maxIndex].Add(time.Second), batch: 1, want: sent[maxIndex].Add(time.Second)},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			count, got, err := m.HighWatermark(ctx, directoryID, logID, tc.start, tc.batch)
+			if err != nil {
+				t.Errorf("HighWatermark(): %v", err)
+			}
+			if !got.Equal(tc.want) {
+				t.Errorf("HighWatermark(%v, %v) high: %v, want %v", tc.start, tc.batch, got, tc.want)
+			}
+			if count != 0 {
+				t.Errorf("HighWatermark(): count %v, want 0", count)
+			}
+		})
+	}
+
+	// Tests that use the watermarks defined during setup.
```

Comment:
It feels like these different test blocks should be different tests. They share the same setup but the blocks are orthogonal.